src/components/QueryBuilder/index.js

```diff
@@ -0,0 +1,298 @@
+  getConfigsByQuery,
+  updateSavedQuery
+} from "../../api/services/configs";
+import { Modal } from "../Modal";
+import { toastError, toastSuccess } from "../Toast/toast";
+import { useLoader } from "../../hooks";
+import { TextInputClearable } from "../TextInputClearable";
+
+export const QueryBuilder = ({ refreshConfigs, className, ...props }) => {
+  const [params, setParams] = useSearchParams();
+  const [query, setQuery] = useState(params.get("query") || "");
+  const [selectedQuery, setSelectedQuery] = useState(null);
+  const [queryList, setQueryList] = useState([]);
+  const [modalIsOpen, setModalIsOpen] = useState(false);
+  const [savedQueryValue, setSavedQueryValue] = useState("");
+  const { loading, setLoading } = useLoader();
+  const options = useMemo(() => {
+    const data = [];
+
+    if (!selectedQuery) {
+      data.push({
+        id: "save",
+        label: "Save",
+        type: "action_save",
+        context: {}
+      });
+    }
+
+    if (selectedQuery) {
+      data.push({
+        id: "update",
+        label: "Update",
+        type: "action_update",
+        context: {}
+      });
+      data.push({
+        id: "save_as",
+        label: "Save As",
+        type: "action_save",
+        context: {}
+      });
+    }
+
+    queryList.forEach((queryItem) => {
+      data.push({
+        id: queryItem.id,
+        label: queryItem.description,
+        type: "action_saved_query",
+        context: {
+          ...queryItem
+        }
+      });
+    });
+    return data;
+  }, [queryList, selectedQuery]);
+
+  const fetchQueries = async () => {
+    setLoading(true);
+    try {
+      const result = await getAllSavedQueries();
+      setQueryList(result?.data || []);
+      setSelectedQuery(null);
+    } catch (ex) {
+      toastError(ex);
+    }
+    setLoading(false);
+  };
+
+  useEffect(() => {
+    fetchQueries();
+    if (params.get("query")) {
+      handleRunQuery();
+    }
+  }, []);
+
+  const handleSearch = (value) => {
+    setQuery(value);
+    setParams({
+      query: value
+    });
+  };
+
+  const handleQueryBuilderAction = (option) => {
+    if (option.type === "action_save") {
+      handleSaveQuery();
+    } else if (option.type === "action_saved_query") {
+      setSelectedQuery(option.context);
+      setQuery(option.context.query);
+    } else if (option.type === "action_update") {
+      updateQuery();
+    }
+  };
+
+  const handleSaveQuery = () => {
+    setSavedQueryValue("");
+    if (!query) {
+      return toastError("Please provide query details");
+    }
+    setModalIsOpen(true);
+  };
+
+  const handleRunQuery = async () => {
+    const queryToRun = query;
+    if (!queryToRun) {
+      toastError("Please provide a query before running");
+      return;
+    }
+    refreshConfigs([]);
+    try {
+      const result = (await getConfigsByQuery(queryToRun))?.data?.results || [];
+      result.forEach((item) => {
+        item.tags = JSON.parse(item.tags);
+      });
```

Comment:
Better to use immutable operations. Easy to mess with mutation. Also, probably better handled inside `getConfigsByQuery`, including the data.results check logic?